### PR TITLE
Include Clustering-Kohonen functions

### PR DESCRIPTION
**Pull request by lorenalves**
Include  functions for clustering using self-organizing maps (as basis kohonen package)

### CODE REVIEW

File: R/sits_kohonen.R [hunk 1]
```diff
@@ -1,92 +1,770 @@
...
+
+    #get label that no have cluster
+    no_cluster<-dplyr::setdiff(temp.data.tb$label, temp.data.tb$neuron_label)
+
+    confusion.matrix<-table(temp.data.tb$label, temp.data.tb$neuron_label)
+
+    #get the names of classes (original labels from samples)
+    Label_table<-rownames(confusion.matrix)
+
+    if (length(no_cluster)>0){
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]: Use space before and after operators in general.

File: R/sits_kohonen.R [hunk 2]
```diff
@@ -1,92 +1,770 @@
...
+
+    #get dimensions (rows and col)
+    #represents the original classes of samples
+    dim_row <-dim(confusion.matrix.tb)[1]
+
+    #represents clusters
+    dim_col <-dim(confusion.matrix.tb)[2]
+
+    #get the names of classes (original labels from samples)
+    Label_table<-rownames(confusion.matrix.tb)[1:dim_row-1]
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]: sits variables and functions always start with lowercase.

File: R/sits_kohonen.R [hunk 3]
```diff
@@ -1,92 +1,770 @@
...
+#' @name .sits_kohonen_pallete
+#' @author Lorena Santos, \email{lorena.santos@@inpe.br}
+#'
+#' @description configure a pallete with 21 colors.
+#'
+#' @return rreturn a pallete of colors
 
-    # create a "cluster" column
-    data.tb$cluster <- as.integer(kohonen.obj$unit.classif)
+.sits_kohonen_pallete <- function()
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]: Is there no color pallete with such colors already implemented in R? R has already a function called `rainbow()`.

File: R/sits_kohonen.R [hunk 4]
```diff
@@ -1,92 +1,770 @@
...
+#' @title Create new groups from kohonen maps
+#' @name sits_subgroup
+#' @author Lorena Santos, \email{lorena.santos@@inpe.br}
+#'
+#' @description Create new groups to identify variations in a same group
+#'
+#' @param koh  An object with informations about kohonen clustering and samples.
+#' @return returns a SITS tibble with subgroups generated by hierarchical clustering.
+#' @export
+sits_subgroup<-function(koh, plot_subgroups = TRUE)
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]: It is never good to have plot and computation in a single function. Better to split it in two (and test them as well).

File: R/sits_kohonen.R [hunk 5]
```diff
@@ -1,92 +1,770 @@
...
+        number_of_cluster<- nb_all$Best.nc[1]
+        number_of_cluster
+
+        distance_atrributes<-proxy::dist(codes_ndvi_evi.ts,  distance = "euclidean")
+        hc<-stats::hclust(distance_atrributes,"ward.D2")
+        cut_hc<-stats::cutree(hc, k = number_of_cluster)
+
+        id_neurons<-names(cut_hc)
+        j = 1
+        #separar os grupos
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]: Do not include comments in portuguese.

File: R/sits_kohonen.R [hunk 6]
```diff
@@ -1,92 +1,770 @@
...
-#' }
+#' @param  ...           additional parameters to be passed to kohonen::supersom function
+#' @return clusters.tb   a tibble with the clusters time series or cluster' members time series according to return_member parameter.
+#' If return_members are FALSE, the returning tibble will contain a new collumn called `n_members` informing how many members has each cluster.
 #' @export
-sits_kohonen <- function (data.tb, bands = NULL, grid_xdim = 5, grid_ydim = 5, rlen = 100,
-                          alpha = c(0.05, 0.01), ...) {
+sits_kohonen <- function (data.tb, time_series, bands = NULL, grid_xdim = 25, grid_ydim = 25, rlen = 100,dist.fcts="euclidean",
+                          alpha = 1,decay.fcts="exponential",neighbourhood.fct ="bubble", ...) {
+
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]: Please define a version name in lorenalves/kohonen that does not match any version of kohonen package in CRAN, and check this version here. If the required version is not the installed version, stop with an error telling the user to install lorenalves/kohonen.

File: R/sits_kohonen.R [hunk 7]
```diff
@@ -1,92 +1,770 @@
...
-#' @author Rolf Simoes, \email{rolf.simoes@@inpe.br}
 #' @author Lorena Alves, \email{lorena.santos@@inpe.br}
 #'
-#' @description This function uses package "kohonen" (self-organized maps) to find clusters in
-#' satellite image time series.
+#' @description This function uses package self-organized maps to find clusters in
+#' satellite image time series to cluster samples.
 #' @references `kohonen` package (https://CRAN.R-project.org/package=kohonen)
 #'
-#' @param data.tb        A tibble with time series to be clustered.
```
[the remark this answers is not in the document]
Comment (lorenalves) [reply R/sits_kohonen.R]: revised

File: R/sits_kohonen.R [hunk 8]
```diff
@@ -1,92 +1,770 @@
...
+#' @param data.tb        a tibble with samples to be clustered
+#' @param time_series    time series extracted from tibble data.
+#' @param bands          bands to be clustered.
+#' @param grid_xdim      x dimension of the SOM grid (default = 5)
+#' @param grid_ydim      y dimension of the SOM grid
+#' @param rlen           number of times the complete data set will be presented to the SOM grid
+#' @param alpha          learning rate, a vector of two numbers indicating the amount of change.
 #'                       Default is to decline linearly from 0.05 to 0.01 over rlen updates.
-#' @param  ...           Additional parameters to be passed to kohonen::supersom function.
-#' @return A full kohonen object for data analysis.
```
[the remark this answers is not in the document]
Comment (lorenalves) [reply R/sits_kohonen.R]: This comment has been changed to adapt to the new kohonen decay function.

File: R/sits_kohonen.R [hunk 9]
```diff
@@ -1,92 +1,770 @@
...
-#' data("samples_MT_9classes")
-#' # compute a kohonen object
-#' koh_obj <- sits_kohonen(samples_MT_9classes)
-#' # generate a dataset
-#' cluster.tb <- sits_kohonen_cluster(samples_MT_9classes, koh_obj)
-#' # plot the result
-#' sits_cluster_frequency(cluster.tb)
-#' }
+#' @param  ...           additional parameters to be passed to kohonen::supersom function
+#' @return clusters.tb   a tibble with the clusters time series or cluster' members time series according to return_member parameter.
```
[the remark this answers is not in the document]
Comment (lorenalves) [reply R/sits_kohonen.R]: revised

File: R/sits_kohonen.R (hunk 2 above)
Comment (lorenalves) [reply R/sits_kohonen.R]: revised

File: R/sits_plot_cluster.R [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        legend = unique(leg[, 1]),
+        col = unique(leg[, 2]),
+        pch = 15,
+        pt.cex = 2,
+        cex = 1,
+        text.col = "black",
+        #horiz = T ,
+        inset = c(0.0350, 0.05),
+        xpd = TRUE,
+        ncol =1
```
Comment (pedro-andrade-inpe) [inline R/sits_plot_cluster.R]:
```suggestion
        ncol = 1
```

File: R/sits_plot_cluster.R [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+#'  \item{"codes" - }{Plot in each neuron the vector weight which corresponds to it }
+#'  \item{"mapping" - }{shows where samples are mapped.}
+#'
+#' }
+#' The sits_plot_kohonen function plots a classified kohonen map.A set of neuron
+#' with same category corresponds to a same cluster.
+#' @param  data  data to be plotted (must be a kohonen object)
+#'
+#' @export
+sits_plot_kohonen <- function(koh, type="codes") {
```
Comment (pedro-andrade-inpe) [inline R/sits_plot_cluster.R]:
```suggestion
sits_plot_kohonen <- function(koh, type = "codes") {
```

File: R/sits_plot_cluster.R [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+    if (type =="mapping"){
```
Comment (pedro-andrade-inpe) [inline R/sits_plot_cluster.R]:
```suggestion
    if (type == "mapping"){
```

File: R/sits_plot_cluster.R [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+#'
+#' @description Plot a bar graph with informations about each cluster.
+#' The percentage of mixture between the clusters.
+#'
+#' @param  data  Table containing the percentage of mixture between the clusters.
+#' @export
+sits_plot_cluster_info <- function(data, text_title = " Cluster ")
+{
+    data <- data$mixture_cluster
+    labels = data$original_class
```
Comment (pedro-andrade-inpe) [inline R/sits_plot_cluster.R]:
```suggestion
    labels <- data$original_class
```

File: R/sits_plot_cluster.R [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+                x = data$cluster,
+                fill = labels
+            ),
+            data = data,
+            stat = "identity",
+            position = position_dodge()
+        )  +
+        theme_minimal() +
+        theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
+        labs(x = "Clusters", y = "Percentage of Mixture", colour ="cluster")+
```
Comment (pedro-andrade-inpe) [inline R/sits_plot_cluster.R]:
```suggestion
        labs(x = "Clusters", y = "Percentage of Mixture", colour = "cluster")+
```

File: R/sits_plot_cluster.R [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+#'
+#' @param  neurons_subgroup  The list contain the EVI and NDVI time series (weight of each neuron) by class.
+#' @export
+sits_plot_subgroups <- function (neurons_subgroup)
+{
+    #get the name of class_neurons from list neurons_subgroup
+    class_neurons <- names(neurons_subgroup)
+    # if (class == "all" )
+    # {
+        i=1
```
Comment (pedro-andrade-inpe) [inline R/sits_plot_cluster.R]:
```suggestion
        i = 1
```

Comment (pedro-andrade-inpe) [reply R/sits_plot_cluster.R]: Review the rest of the code.

File: R/sits_kohonen.R [hunk 16]
```diff
@@ -1,92 +1,695 @@
...
 #' @references `kohonen` package (https://CRAN.R-project.org/package=kohonen)
 #'
-#' @param data.tb        A tibble with time series to be clustered.
+#' @param data.tb        A tibble with samples to be clustered.
+#' @param time_series    Time series extracted from tibble data.
 #' @param bands          Bands to be clustered.
 #' @param grid_xdim      X dimension of the SOM grid (default = 5).
 #' @param grid_ydim      Y dimension of the SOM grid.
-#' @param rlen           Number of times the complete data set will be presented to the SOM grid.
+#' @param rlen           Number of times the complete data set will be presented to the SOM grid
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]:
```suggestion
#' @param rlen           Number of times the complete data set will be presented to the SOM grid.
```

File: R/sits_kohonen.R [hunk 17]
```diff
@@ -1,92 +1,695 @@
...
-#' sits_cluster_frequency(cluster.tb)
-#' }
+#' @return  A tibble with the clusters time series or cluster' members time series according to return_member parameter.
+#' If return_members are FALSE, the returning tibble will contain a new collumn called `n_members` informing how many members has each cluster.
 #' @export
-sits_kohonen <- function (data.tb, bands = NULL, grid_xdim = 5, grid_ydim = 5, rlen = 100,
-                          alpha = c(0.05, 0.01), ...) {
-    # does the input data exist?
-    .sits_test_tibble(data.tb)
+sits_kohonen <- function (data.tb, time_series, bands = NULL, grid_xdim = 25, grid_ydim = 25, rlen = 100,dist.fcts="euclidean",
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]:
```suggestion
sits_kohonen <- function (data.tb, time_series, bands = NULL, grid_xdim = 25, grid_ydim = 25, rlen = 100, dist.fcts = "euclidean",
```

File: R/sits_kohonen.R [hunk 18]
```diff
@@ -1,92 +1,770 @@
...
+                          alpha = 1,decay.fcts="exponential",neighbourhood.fct ="bubble", ...) {
+
+
+    palleteVizin <- .sits_kohonen_pallete()
     # does the input data exist?
-    .sits_test_tibble(data.tb)
+    .sits_test_tibble (data.tb)
 
-    # if no bands informed, get all bands available in sits tibble
+    # if no bands informed, get all bands available in SITS tibble
```
[the remark this answers is not in the document]
Comment (pedro-andrade-inpe) [reply R/sits_kohonen.R]: This issue was not solved.

File: R/sits_kohonen.R [hunk 19]
```diff
@@ -1,92 +1,695 @@
...
+    table_class_matrix_id <- (unique(class_matrix))
+
+    #Add at the sample the label of neuron
+    neurons_ <- result.tb$id_neuron
+
+    # Which cluster the samples was allocated?
+    cluster_sample <- class_matrix[neurons_]
+    result.tb$neuron_label <- cluster_sample
+
+    j=1
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]:
```suggestion
    j = 1
```

File: R/sits_kohonen.R [hunk 20]
```diff
@@ -1,92 +1,695 @@
...
-    # return kohonen object
-    return (kohonen.obj)
+
+    kohonen_obj$paint_map <- neighborhood
+    kohonen_obj$neurons_labelled <- neurons_labelled
+
+
+    info_samples_tables <- structure(list(kohonen_obj = kohonen_obj,
+                                          info_samples = result.tb),
+                                     class = "sits")
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]: There is an indentation problem here.

File: R/sits_kohonen.R [hunk 21]
```diff
@@ -1,92 +1,695 @@
...
+sits_evaluate_samples <- function(data.tb,
+                                time_series,
+                                grid_xdim = 5,
+                                grid_ydim = 5,
+                                rlen = 100,
+                                alpha = c(1),
+                                radius = 6,
+                                distance = "euclidean",
+                                iterations = 1,
+                                mode="online")
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]:
```suggestion
                                mode = "online")
```

File: R/sits_kohonen.R [hunk 22]
```diff
@@ -1,92 +1,695 @@
...
+#' @description Labelling all neurons using the majority voting.
+#' If the neuron is empty, It will labeled as "NoClass".
+#'
+#' @param data.tb    A SITS tibble with info of samples and kohonen.obj.
+#' @param size_grid  The size of kohonen map.
+#' @return returns the labels of each neuron.
+#'
+.sits_labelling_neurons <- function (data.tb, grid_size)
+{
+    i=1
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]: This line is unnecessary.

File: R/sits_kohonen.R [hunk 23]
```diff
@@ -1,92 +1,695 @@
...
+        dplyr::select(info_sample_cluster.tb,
+                      id_sample,
+                      original_label,
+                      neuron_label,
+                      frequency,
+                      percentage,
+                      cluster_label)
+
+    #join samples with new cluster
+    info_samples_id_cluster<-unique(dplyr::select(info_sample_cluster.tb, id_sample, cluster_label))
```
Comment (pedro-andrade-inpe) [inline R/sits_kohonen.R]:
```suggestion
    info_samples_id_cluster <- unique(dplyr::select(info_sample_cluster.tb, id_sample, cluster_label))
```

**Pull request merged**